Пример на датасете из репозитория UCI
ILPD (Indian Liver Patient Dataset)
Описание данных - https://archive.ics.uci.edu/ml/datasets/ILPD+%28Indian+Liver+Patient+Dataset%29

In [89]:
import pandas as pd
import numpy as np
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
data = pd.read_csv("1.csv", header=None)
data.head(13)

0       1     2    3    4   5    6    7    8     9   10
0   65  Female   0.7  0.1  187  16   18  6.8  3.3  0.90   1
1   62    Male  10.9  5.5  699  64  100  7.5  3.2  0.74   1
2   62    Male   7.3  4.1  490  60   68  7.0  3.3  0.89   1
3   58    Male   1.0  0.4  182  14   20  6.8  3.4  1.00   1
4   72    Male   3.9  2.0  195  27   59  7.3  2.4  0.40   1
5   46    Male   1.8  0.7  208  19   14  7.6  4.4  1.30   1
6   26  Female   0.9  0.2  154  16   12  7.0  3.5  1.00   1
7   29  Female   0.9  0.3  202  14   11  6.7  3.6  1.10   1
8   17    Male   0.9  0.3  202  22   19  7.4  4.1  1.20   2
9   55    Male   0.7  0.2  290  53   58  6.8  3.4  1.00   1
10  57    Male   0.6  0.1  210  51   59  5.9  2.7  0.80   1
11  72    Male   2.7  1.3  260  31   56  7.4  3.0  0.60   1
12  64    Male   0.9  0.3  310  61   58  7.0  3.4  0.90   2

У нас есть 10 признаков и 1 целевая переменная (бинарная) - нужно определить имеется болезнь или нет

In [60]:
print(data.shape)

(582, 11)


Всего 582 случая

Посмотрим на соотношение классов

In [61]:
data.iloc[:, -1].value_counts()

1    415
2    167
Name: 1, dtype: int64

In [90]:
#data['Female'] = data['Female'].astype('category')
dct = {1: 1, 2: 0}
data[10] = data[10].map(dct)
data.head(13)
#data.drop('Female', axis=1)

0       1     2    3    4   5    6    7    8     9   10
0   65  Female   0.7  0.1  187  16   18  6.8  3.3  0.90   1
1   62    Male  10.9  5.5  699  64  100  7.5  3.2  0.74   1
2   62    Male   7.3  4.1  490  60   68  7.0  3.3  0.89   1
3   58    Male   1.0  0.4  182  14   20  6.8  3.4  1.00   1
4   72    Male   3.9  2.0  195  27   59  7.3  2.4  0.40   1
5   46    Male   1.8  0.7  208  19   14  7.6  4.4  1.30   1
6   26  Female   0.9  0.2  154  16   12  7.0  3.5  1.00   1
7   29  Female   0.9  0.3  202  14   11  6.7  3.6  1.10   1
8   17    Male   0.9  0.3  202  22   19  7.4  4.1  1.20   0
9   55    Male   0.7  0.2  290  53   58  6.8  3.4  1.00   1
10  57    Male   0.6  0.1  210  51   59  5.9  2.7  0.80   1
11  72    Male   2.7  1.3  260  31   56  7.4  3.0  0.60   1
12  64    Male   0.9  0.3  310  61   58  7.0  3.4  0.90   0

In [86]:
data[10]

0      1
1      1
2      1
3      1
4      1
      ..
578    2
579    1
580    1
581    1
582    2
Name: 10, Length: 583, dtype: int64

In [93]:
cat_attribs = [1]
full_pipeline = ColumnTransformer([('cat', OneHotEncoder(handle_unknown='ignore'), cat_attribs)], remainder='passthrough')

encoder = full_pipeline.fit(data)
data = encoder.transform(data)

In [94]:
data = pd.DataFrame(data)

In [95]:
data.head(13)

0    1     2     3    4      5     6      7    8    9     10   11
0   1.0  0.0  65.0   0.7  0.1  187.0  16.0   18.0  6.8  3.3  0.90  1.0
1   0.0  1.0  62.0  10.9  5.5  699.0  64.0  100.0  7.5  3.2  0.74  1.0
2   0.0  1.0  62.0   7.3  4.1  490.0  60.0   68.0  7.0  3.3  0.89  1.0
3   0.0  1.0  58.0   1.0  0.4  182.0  14.0   20.0  6.8  3.4  1.00  1.0
4   0.0  1.0  72.0   3.9  2.0  195.0  27.0   59.0  7.3  2.4  0.40  1.0
5   0.0  1.0  46.0   1.8  0.7  208.0  19.0   14.0  7.6  4.4  1.30  1.0
6   1.0  0.0  26.0   0.9  0.2  154.0  16.0   12.0  7.0  3.5  1.00  1.0
7   1.0  0.0  29.0   0.9  0.3  202.0  14.0   11.0  6.7  3.6  1.10  1.0
8   0.0  1.0  17.0   0.9  0.3  202.0  22.0   19.0  7.4  4.1  1.20  0.0
9   0.0  1.0  55.0   0.7  0.2  290.0  53.0   58.0  6.8  3.4  1.00  1.0
10  0.0  1.0  57.0   0.6  0.1  210.0  51.0   59.0  5.9  2.7  0.80  1.0
11  0.0  1.0  72.0   2.7  1.3  260.0  31.0   56.0  7.4  3.0  0.60  1.0
12  0.0  1.0  64.0   0.9  0.3  310.0  61.0   58.0  7.0  3.4  0.90  0.0

Разбиваем выборку на тренировочную и тестовую части и обучаем модель (в примере - градиентный бустинг)

In [96]:
from sklearn.model_selection import train_test_split

x_data = data.iloc[:,:-1]
y_data = data.iloc[:,-1]

x_train, x_test, y_train, y_test = train_test_split(x_data, y_data, test_size=0.2, random_state=7)

In [97]:
import xgboost as xgb

model = xgb.XGBClassifier()

model.fit(x_train, y_train)
y_predict = model.predict(x_test)

[16:29:59] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


C:\Users\serva\anaconda3\lib\site-packages\xgboost\sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


Проверяем качество

In [98]:
from sklearn.metrics import recall_score, precision_score, roc_auc_score, accuracy_score, f1_score

def evaluate_results(y_test, y_predict):
    print('Classification results:')
    f1 = f1_score(y_test, y_predict)
    print("f1: %.2f%%" % (f1 * 100.0)) 
    roc = roc_auc_score(y_test, y_predict)
    print("roc: %.2f%%" % (roc * 100.0)) 
    rec = recall_score(y_test, y_predict, average='binary')
    print("recall: %.2f%%" % (rec * 100.0)) 
    prc = precision_score(y_test, y_predict, average='binary')
    print("precision: %.2f%%" % (prc * 100.0)) 

    
evaluate_results(y_test, y_predict)

Classification results:
f1: 82.02%
roc: 61.73%
recall: 90.12%
precision: 75.26%


### Теперь очередь за PU learning

Представим, что нам неизвестны негативы и часть позитивов

In [193]:
mod_data = data.copy()
#get the indices of the positives samples
pos_ind = np.where(mod_data.iloc[:,-1].values == 1)[0]
#shuffle them
np.random.shuffle(pos_ind)
# leave just 25% of the positives marked
pos_sample_len = int(np.ceil(0.05 * len(pos_ind)))
print(f'Using {pos_sample_len}/{len(pos_ind)} as positives and unlabeling the rest')
pos_sample = pos_ind[:pos_sample_len]

Using 21/416 as positives and unlabeling the rest


Создаем столбец для новой целевой переменной, где у нас два класса - P (1) и U (-1)

In [194]:
mod_data['class_test'] = -1
mod_data.loc[pos_sample,'class_test'] = 1
print('target variable:\n', mod_data.iloc[:,-1].value_counts())

target variable:
 -1    562
 1     21
Name: class_test, dtype: int64


In [195]:
mod_data.head(10)

0    1     2     3    4      5     6      7    8    9    10   11  \
0  1.0  0.0  65.0   0.7  0.1  187.0  16.0   18.0  6.8  3.3  0.90  1.0   
1  0.0  1.0  62.0  10.9  5.5  699.0  64.0  100.0  7.5  3.2  0.74  1.0   
2  0.0  1.0  62.0   7.3  4.1  490.0  60.0   68.0  7.0  3.3  0.89  1.0   
3  0.0  1.0  58.0   1.0  0.4  182.0  14.0   20.0  6.8  3.4  1.00  1.0   
4  0.0  1.0  72.0   3.9  2.0  195.0  27.0   59.0  7.3  2.4  0.40  1.0   
5  0.0  1.0  46.0   1.8  0.7  208.0  19.0   14.0  7.6  4.4  1.30  1.0   
6  1.0  0.0  26.0   0.9  0.2  154.0  16.0   12.0  7.0  3.5  1.00  1.0   
7  1.0  0.0  29.0   0.9  0.3  202.0  14.0   11.0  6.7  3.6  1.10  1.0   
8  0.0  1.0  17.0   0.9  0.3  202.0  22.0   19.0  7.4  4.1  1.20  0.0   
9  0.0  1.0  55.0   0.7  0.2  290.0  53.0   58.0  6.8  3.4  1.00  1.0   

   class_test  
0          -1  
1          -1  
2          -1  
3          -1  
4          -1  
5          -1  
6          -1  
7          -1  
8          -1  
9          -1

Remember that this data frame (x_data) includes the former target variable that we keep here just to compare the results

[:-2] is the original class label for positive and negative data [:-1] is the new class for positive and unlabeled data

In [196]:
x_data = mod_data.iloc[:,:-2].values # just the X 
y_labeled = mod_data.iloc[:,-1].values # new class (just the P & U)
y_positive = mod_data.iloc[:,-2].values # original class

### 1. random negative sampling

In [197]:
mod_data = mod_data.sample(frac=1)
neg_sample = mod_data[mod_data['class_test']==-1][:len(mod_data[mod_data['class_test']==1])]
sample_test = mod_data[mod_data['class_test']==-1][len(mod_data[mod_data['class_test']==1]):]
pos_sample = mod_data[mod_data['class_test']==1]
print(neg_sample.shape, pos_sample.shape)
sample_train = pd.concat([neg_sample, pos_sample]).sample(frac=1)

(21, 13) (21, 13)


In [198]:
model = xgb.XGBClassifier()

model.fit(sample_train.iloc[:,:-2].values, 
          sample_train.iloc[:,-2].values)
y_predict = model.predict(sample_test.iloc[:,:-2].values)
evaluate_results(sample_test.iloc[:,-2].values, y_predict)

[16:43:50] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
Classification results:
f1: 81.93%
roc: 55.54%
recall: 93.68%
precision: 72.80%


C:\Users\serva\anaconda3\lib\site-packages\xgboost\sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


### Домашнее задание

1. взять любой набор данных для бинарной классификации (можно скачать один из модельных с https://archive.ics.uci.edu/ml/datasets.php)
3. сделать feature engineering
4. обучить любой классификатор (какой вам нравится)
5. далее разделить ваш набор данных на два множества: P (positives) и U (unlabeled). Причем брать нужно не все положительные (класс 1) примеры, а только лишь часть
6. применить random negative sampling для построения классификатора в новых условиях
7. сравнить качество с решением из пункта 4 (построить отчет - таблицу метрик)
8. поэкспериментировать с долей P на шаге 5 (как будет меняться качество модели при уменьшении/увеличении размера P)

<b>Бонусный вопрос:</b>

Как вы думаете, какой из методов на практике является более предпочтительным: random negative sampling или 2-step approach?

Ваш ответ здесь:

Доля 25%
Classification results 25%:
f1: 77.78%
roc: 54.93%
recall: 92.56%
precision: 67.07%
    
Доля 5%
Classification results:
f1: 80.62%
roc: 57.35%
recall: 88.45%
precision: 74.07%
    
Доля 20%
Classification results:
f1: 80.37%
roc: 55.95%
recall: 93.24%
precision: 70.62%
    
Доля 31%
Classification results:
f1: 76.80%
roc: 56.31%
recall: 92.12%
precision: 65.85%
    
При сильном увлеичении или уменьшении доли P может изменяться в лучшую сторону показатель recall, при этом precision падает до 50%, что в случае определения болезни не подходит.
